Packages/com.unity.inputsystem/InputSystem/Events/InputEventTrace.cs

```diff
@@ -1,85 +1,648 @@
+        /// </summary>
+        /// <value>Total size of event data currently in trace.</value>
+        public long totalEventSizeInBytes => m_EventSizeInBytes;
+
+        /// <summary>
+        /// Total size of memory buffer (in bytes) currently allocated.
+        /// </summary>
+        /// <value>Size of memory currently allocated.</value>
+        /// <remarks>
+        /// The buffer is allocated on the unmanaged heap.
+        /// </remarks>
+        public long allocatedSizeInBytes => m_EventBuffer != default ? m_EventBufferSize : 0;
+
+        /// <summary>
+        /// Largest size (in bytes) that the memory buffer is allowed to grow to. By default, this is
+        /// the same as <see cref="allocatedSizeInBytes"/> meaning that the buffer is not allowed to grow but will
+        /// rather wrap around when full.
+        /// </summary>
+        /// <value>Largest size the memory buffer is allowed to grow to.</value>
+        public long maxSizeInBytes => m_MaxEventBufferSize;
+
+        /// <summary>
+        /// Information about all devices for which events have been recorded in the trace.
+        /// </summary>
+        /// <value>Record of devices recorded in the trace.</value>
+        public ReadOnlyArray<DeviceInfo> deviceInfos => m_DeviceInfos;
+
+        /// <summary>
+        /// Optional delegate to decide whether an input should be stored in a trace. Null by default.
+        /// </summary>
+        /// <value>Delegate to accept or reject individual events.</value>
+        /// <remarks>
+        /// When this is set, the callback will be invoked on every event that would otherwise be stored
+        /// directly in the trace. If the callback returns <c>true</c>, the trace will continue to record
+        /// the event. If the callback returns <c>false</c>, the event will be ignored and not recorded.
+        ///
+        /// The callback should generally mutate the event. If you do so, note that this will impact
+        /// event processing in general, not just recording of the event in the trace.
+        /// </remarks>
+        public Func<InputEventPtr, InputDevice, bool> onFilterEvent
+        {
+            get => m_OnFilterEvent;
+            set => m_OnFilterEvent = value;
+        }
+
+        /// <summary>
+        /// Event that is triggered every time an event has been recorded in the trace.
+        /// </summary>
         public event Action<InputEventPtr> onEvent
         {
-            add => m_EventListeners.Append(value);
+            add
+            {
+                if (!m_EventListeners.ContainsReference(value))
+                    m_EventListeners.Append(value);
+            }
             remove => m_EventListeners.Remove(value);
         }
 
-        // Create a disabled event trace that does not perform any allocation
-        // yet. An event trace only starts consuming resources the first time
-        // it is enabled.
-        public InputEventTrace(int bufferSize = kDefaultBufferSize)
```

Comment:
FYI: Removing this constructor breaks assembly compatibility. You still have source compatibility, as there is the constructor which takes buffer size and more parameters, which all have default values, which the compiler can then map to. But that's a compile-time mapping, not runtime. So any dll referencing the old constructor would fail. I don't particularly care because:
1. Package still has the preview label
2. Forward/Backwards compatibility of assemblies referencing packages is grey area in Unity anyways. It is not something we really want to commit to supporting, because it is a mess, but people do it all the time anyways.

Reply:
> Changes lgtm (except for the #ifdefed out code) - though I'm wary of the amount of changes landing in 1.0.0-preview.x at this point. I would suggest a branch-off for any new work soon. Please have someone do some manual QA on this before landing, though!

Asked Stefan for QA validation.

Reply:
(wtf... no idea why GitHub put the response here...)

Reply:
Stefan uncovered a couple cosmetic issues. Fixed. Should be ready to go.